File: Programação Orientada a Objetos/JogoTabuleiro/tabuleiro.java
public class tabuleiro {
    private int largura;
    private int comprimento;
    private int zero_fileira;
    private int zero_coluna;
    private int[][] tbl;

    public tabuleiro(int largura, int comprimento) {
        this.largura = largura;
        this.comprimento = comprimento;
        tbl = new int[comprimento][largura];
    }

    public void preencher_tabuleiro(int[] valores, int quantidade) {
        if(quantidade == comprimento*largura) {
            for(int i = 0; i < comprimento; i++) 
                for(int j = 0; j < largura; j++) {
                    tbl[i][j] = valores[largura*i + j];
                    if(valores[largura*i + j] == 0) {
                        zero_fileira = i;
                        zero_coluna = j;
                    }
                } 
        }
    }

    private void swap(int i1, int j1, int i2, int j2) {
        int aux = tbl[i1][j1];
        tbl[i1][j1] = tbl[i2][j2];
        tbl[i2][j2] = aux;
    } 

    private void tabuleiro_up() {
        if(zero_fileira < comprimento - 1) {
            swap(zero_fileira + 1, zero_coluna, zero_fileira, zero_coluna);
            zero_fileira += 1;
        }
    }
    private void tabuleiro_down() {
        if(zero_fileira > 0) {
            swap(zero_fileira - 1, zero_coluna, zero_fileira, zero_coluna);
            zero_fileira -= 1;
        }
    }

    private void tabuleiro_left() {
        if(zero_coluna < largura - 1) {
            swap(zero_fileira, zero_coluna + 1, zero_fileira, zero_coluna);
            zero_coluna += 1;
        }
    }

    private void tabuleiro_right() {
        if(zero_coluna > 0) {
            swap(zero_fileira, zero_coluna - 1, zero_fileira, zero_coluna);
            zero_coluna -= 1;
        }
    }


    public void movimento(char c) {
        switch (c) {
            case 'u':
                tabuleiro_up();
                break;
        
            case 'd':
                tabuleiro_down();
                break;
    
            case 'l':
                tabuleiro_left();
                break;
        
            case 'r':
                tabuleiro_right();
                break; 
        }
    }

    private void print_line(int tamanho) {
        for(int i = 0; i < tamanho; i++) {
            System.out.print("+------");
        }
        System.out.println("+");
    }

    
    public void printa_tabuleiro() {
        print_line(largura);
        
        for(int i = 0; i < comprimento; i++) {
            System.out.print("|");
            for(int j = 0; j < largura; j++) {
                if(tbl[i][j] >= 10)
                    System.out.print("  " + tbl[i][j] + "  |");
                else
                    System.out.print("   " + tbl[i][j] + "  |");
            }

            System.out.println();
            print_line(largura);
        }

        System.out.println();
    }


    public boolean Ehsolucao() {
        for(int i = 0; i < comprimento; i++)
            for(int j = 0; j < largura; j++)
                if((tbl[i][j]) != largura*i + j)
                    return false;

        return(true);
    }



}

File: Programação Orientada a Objetos/JogoTabuleiroPython/tabuleiro.py
import math

class Tabuleiro:

    def __init__(self, comprimento, largura):
        self.comprimento = comprimento
        self.largura = largura
        self.tabuleiro = []
        self.pos_vazia_i = -1
        self.pos_vazia_j = -1

    def preenche_tabuleiro(self, valores):
        self.largura =  int(math.sqrt(len(valores)))
        self.comprimento =  int(math.sqrt(len(valores)))
        self.tabuleiro = [[0 for i in range(self.largura)] for x in range(self.comprimento)]
       
        for i in range(self.comprimento):
            for j in range(self.largura):
                numero = valores[self.largura*i + j]
                self.tabuleiro[i][j] = numero
                if(numero == 0):
                    self.pos_vazia_i = i
                    self.pos_vazia_j = j

    def swap(self, i1, j1, i2, j2):
        tmp = self.tabuleiro[i1][j1]
        self.tabuleiro[i1][j1] = self.tabuleiro[i2][j2]
        self.tabuleiro[i2][j2] = tmp

    
    def up(self):
        if(self.pos_vazia_i < self.comprimento - 1 and self.pos_vazia_i != -1):
            self.swap(self.pos_vazia_i + 1, self.pos_vazia_j, self.pos_vazia_i, self.pos_vazia_j)
            self.pos_vazia_i += 1

    def down(self):
        if(self.pos_vazia_i > 0):
            self.swap(self.pos_vazia_i - 1, self.pos_vazia_j, self.pos_vazia_i, self.pos_vazia_j)
            self.pos_vazia_i -= 1

    def left(self):
        if(self.pos_vazia_j < self.largura - 1  and  self.pos_vazia_j != -1):
            self.swap(self.pos_vazia_i, self.pos_vazia_j + 1, self.pos_vazia_i, self.pos_vazia_j)
            self.pos_vazia_j += 1

    def right(self):
        if(self.pos_vazia_j > 0):
            self.swap(self.pos_vazia_i, self.pos_vazia_j - 1, self.pos_vazia_i, self.pos_vazia_j)
            self.pos_vazia_j -= 1


    def movements(self):
        movements = input()
        self.printa_tabuleiro()
        for i in range(len(movements)):
            match movements[i]:
                case 'u':
                    self.up()
                case 'd':
                    self.down()
                case 'l':
                    self.left()
                case 'r':
                    self.right()

            self.printa_tabuleiro()


    def print_line(self, tamanho):
        for i in range(tamanho):
            print("+------", end = "")

        print("+")

    def printa_tabuleiro(self):
        self.print_line(self.largura)

        for i in range(self.comprimento):
            print("|", end = "")
            for j in range(self.largura):
                if(self.tabuleiro[i][j] >= 10):
                    print(" ", self.tabuleiro[i][j], end = "")
                    print("  |", end = "")
                else:
                    print("   ", end = "")
                    if(self.tabuleiro[i][j] != 0):
                        print(self.tabuleiro[i][j], end = "")
                    else:
                        print(" ", end = "")          
                    print("  |", end = "")

            print()
            self.print_line(self.largura)

        print()


    def EhSolucao(self):
        for i in range(self.comprimento):
            for j in range(self.largura):
                if(self.tabuleiro[i][j] != self.largura*i + j):
                    return False 

        return True




def leitura():
    entrada = input()
    inteiros = entrada.split()
    for i in range(len(inteiros)):
        inteiros[i] = int(inteiros[i])
    return inteiros


def main():
    valores = leitura()
    tabuleiro = Tabuleiro(0, 0) 

    tabuleiro.preenche_tabuleiro(valores)
    tabuleiro.movements()
    if(tabuleiro.EhSolucao()):
        print("Posicao final: True")
    else:
        print("Posicao final: False")


main()